### PR TITLE
[libc++][hardening] In production hardening modes, trap rather than abort

### PR DESCRIPTION
**Pull request by var-const**
In the hardening modes that can be used in production (`fast` and `extensive`), make a failed assertion invoke a trap instruction rather than calling verbose abort. In the debug mode, still keep calling verbose abort to provide a better user experience and to allow us to keep our existing testing infrastructure for verifying assertion messages. Since the debug mode by definition enables all assertions, we can be sure that we still check all the assertion messages in the library when running the test suite in the debug mode.

The main motivation to use trapping in production is to achieve better code generation and reduce the binary size penalty. This way, the assertion handler can compile to a single instruction, whereas the existing mechanism with verbose abort results in generating a function call that in general cannot be optimized away (made worse by the fact that it's a variadic function, imposing an additional penalty). See the [RFC](https://discourse.llvm.org/t/rfc-hardening-in-libc/73925) for more details. Note that this mechanism can now be completely [overridden at CMake configuration time](https://github.com/llvm/llvm-project/pull/77883).

This patch also significantly refactors `check_assertion.h` and expands its test coverage. The main changes:
- when overriding `verbose_abort`, don't do matching inside the function -- just print the error message to `stderr`. This removes the need to set a global matcher and allows to do matching in the parent process after the child finishes;
- remove unused logic for matching source locations and for using wildcards;
- make matchers simple functors;
- introduce `DeathTestResult` that keeps data about the test run, primarily to make it easier to test.

In addition to the refactoring, `check_assertion.h` can now recognize when a process exits due to a trap.

### CODE REVIEW

File: libcxx/test/support/test.support/test_check_assertion.pass.cpp
Comment (ldionne) [inline libcxx/test/support/test.support/test_check_assertion.pass.cpp:107]: Could you also test the macros we define in `check_assertion.h`? `EXPECT_DEATH` & friends should have some coverage here.

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/vendor/llvm/default_assertion_handler.in
Comment (ldionne) [inline libcxx/vendor/llvm/default_assertion_handler.in:20]: As you mentioned, this needs a release note and probably documentation updates too.

File: libcxx/test/support/check_assertion.h
Comment (var-const) [reply libcxx/test/support/check_assertion.h:248]: I like this!

File: libcxx/test/support/check_assertion.h
Comment (var-const) [reply libcxx/test/support/check_assertion.h:128]: At this point, I can't remember -- evidently, I wanted to allow partial matches for some reason. Looking at the original code, it's also confusing -- it seems to be just doing `==` in an unnecessarily verbose way. Replaced with a simple equality comparison.

File: libcxx/test/support/check_assertion.h
Comment (var-const) [reply libcxx/test/support/check_assertion.h:263]: Done. I also couldn't resist the temptation to remove all the wildcard matching logic and matching files and lines -- this isn't currently used anywhere (perhaps it was used at some point before?), and since it's non-trivial logic that also isn't tested, IMO leaving it in place does more harm than good.

File: libcxx/test/support/check_assertion.h
Comment (var-const) [reply libcxx/test/support/check_assertion.h:128]: (I think I remember -- I probably wanted to be able to match the message without parsing the full regex format)

File: libcxx/docs/ReleaseNotes/18.rst
Comment (var-const) [inline libcxx/docs/ReleaseNotes/18.rst:114]: Tagging @llvm/libcxx-vendors

File: libcxx/docs/ReleaseNotes/18.rst
Comment (ldionne) [inline libcxx/docs/ReleaseNotes/18.rst:148]:
```suggestion
- The ability to override `__libcpp_verbose_abort` will be removed in an upcoming release.
```

We'll still keep the function but it's an implementation detail.

File: libcxx/test/support/test.support/test_check_assertion.pass.cpp
Comment (ldionne) [inline libcxx/test/support/test.support/test_check_assertion.pass.cpp:89]: Leftover?

File: libcxx/docs/BuildingLibcxx.rst
Comment (ldionne) [inline libcxx/docs/BuildingLibcxx.rst:498]:
```suggestion
Under the hood, a hardening assertion will invoke the ``_LIBCPP_ASSERTION_HANDLER`` macro upon failure.
A vendor may provide a header that contains
a custom definition of this macro and specify the path to the header via the
```

File: libcxx/docs/BuildingLibcxx.rst
Comment (ldionne) [inline libcxx/docs/BuildingLibcxx.rst:494]:
```suggestion
``std::abort()``). This is done to minimize the code size impact of enabling hardening in the library. However, vendors can also override that mechanism at CMake
configuration time.
```

**Review by ldionne — APPROVED** [review]
This LGTM with minor comments, thanks! This makes us very close to the final state explained in the hardening RFC! I think the main thing that is left is to switch to `__builtin_verbose_trap(message)` once that is implemented.

**Comment by var-const** [comment]
For the record, with the changes applied, the output produced by `check_assertion.h` looks like this:
```
# .---command stderr------------
# | Failure: EXPECT_DEATH( c.front() ) failed!
# | (cause: Child died, but with a different death cause
# | Expected cause:   verbose abort
# | Actual cause:     trap
# | )
# |
# | child exit code: 5
# | ---------- standard err ----------
# |
# | ----------------------------------
# | ---------- standard out ----------
# |
# | ----------------------------------
# | Assertion failed: (( ExpectDeath(DeathCause::VerboseAbort, "c.front()", [&]() { (void)(c.front()); }) )), function main, file assert.cfront.empty.pass.cpp, line 28.
# `-----------------------------
# error: command failed with exit status: 250
```

```
# .---command stderr------------
# | Failure: EXPECT_DEATH( c.front() ) failed!
# | (cause: Child died, but with a different verbose abort message
# | Expected message:   'front() called on an empty container'
# | Actual message:     'front() called on an empty vector'
# | Expected location:   *:*
# | Actual location:     /Users/varconst/work/llvm-project/build/include/c++/v1/vector:617
# | )
# |
# | child exit code: 1
# | ---------- standard err ----------
# | ###
# | /Users/varconst/work/llvm-project/build/include/c++/v1/vector:617: assertion !empty() failed: front() called on an empty vector
# | ###
# | ----------------------------------
# | ---------- standard out ----------
# |
# | ----------------------------------
# | Assertion failed: (( ExpectDeath(DeathCause::VerboseAbort, "c.front()", [&]() { (void)(c.front()); }, AssertionInfoMatcher("front() called on an empty container")) )), function main, file assert.cfront.empty.pass.cpp, line 35.
# `-----------------------------
# error: command failed with exit status: 250
```

**Comment by var-const** [comment]
Note on the CI before merging:
1. The few stage 3 failures are spurious (cancelled jobs for `bootstrapping-build`, `generic-msan` and `no-tzdb`). There is no reason to suspect that this jobs might not succeed, and they did succeed on a previous version of this patch.
2. The `picolibc` failures on [Buildkite](https://buildkite.com/llvm-project/libcxx-ci/builds/33012#_) (the `ARM` jobs) are known issues and unrelated.
3. The Windows and FreeBSD jobs on [Buildkite](https://buildkite.com/llvm-project/libcxx-ci/builds/33012#_) have been still waiting for an agent at the time of merge. These jobs are currently heavily backlogged. There is no particular reason to suspect this would fail on those platforms. I will proceed with the merge, making sure to address any problems should those jobs eventually fail.

File: libcxx/test/libcxx/algorithms/alg.sorting/assert.sort.invalid_comparator.pass.cpp
Comment (var-const) [inline libcxx/test/libcxx/algorithms/alg.sorting/assert.sort.invalid_comparator.pass.cpp:14]: Note: while for tests that already require the debug mode this is, strictly speaking, redundant, I think that a) it's slightly easier to make this condition uniform throughout the test suite; and b) this is slightly more robust should that requirement ever be lowered (e.g. if changing this test started to run in the extensive mode, it's easy to forget to update the `XFAIL` condition, which might then take a long time to show up on the CI or might even not be covered by the CI).

**Comment by vitalybuka** [comment]
Breaks with sanitizers https://lab.llvm.org/buildbot/#/builders/236/builds/8851/steps/11/logs/stdio

**Comment by var-const** [comment]
> Breaks with sanitizers https://lab.llvm.org/buildbot/#/builders/236/builds/8851/steps/11/logs/stdio

@vitalybuka Does this only break with HWAsan, or with other sanitizers as well?

**Comment by vitalybuka** [comment]
Only HWASAN, and only HWASAN intercept TRAP by default

**Pull request merged**
